README.md
# Locationator

A simple macOS menubar app that provides access to the macOS Location Services reverse geocoding API. Also includes a command line tool for accessing the reverse geocoding API and for writing reverse geocoding data to the XMP metadata of image and video files as well as two services for accessing the reverse geocoding API from the Services menu.

## Rationale

Why this app? Apple's Location Services API is great and I wanted to be able to use the reverse geocoding API from the command line. Unfortunately, in Ventura, Apple has basically made it impossible to use Location Services from a command line app because the app cannot be granted the necessary permissions in System Settings. This app is a workaround for that problem. It creates a menubar app that can be granted the necessary permissions and then provides a simple API for accessing the reverse geocoding API via a built in web server. This means that you can use the Location Services API from the command line by making a simple HTTP request to the built in web server via curl for example or can use the API end point from any other app that can make HTTP requests.

## Installation

Download the latest release DMG from the [releases page](https://github.com/RhetTbull/locationator/releases) and use the DMG to install the app. Once installed, launch the app and grant it the necessary permissions.

To launch Locationator the first time you'll need to right-click on the app icon and select "Open" otherwise you may get a warning about unknown developer as the app is not signed with an Apple Developer ID. You may need to do this twice.

Alternatively, to build from source:

- clone the repo
- cd into the repo directory
- create a virtual environment and activate it
- python3 -m pip install -r requirements.txt
- python3 setup.py py2app
- Copy dist/Locationator.app to /Applications

## Screenshots

![Locationator](images/menubar.png)

![Locationator](images/reverse_geocode.png)

## Usage

Locationator server is a very simple HTTP server for handling local requests. It supports two endpoints, `GET /` and `GET /reverse_geocode`.

>*Please note*, this server is for local use and NOT intended to be exposed to the internet. The server does not support any authentication or authorization and is intended to be used on a local machine only.

### GET /

This endpoint provides the current version and the port on which the server is running.

**URL** : `/`

**Method** : `GET`

**Response format** : Content-type is plain text

**Success Response Example** :

*Note*: examples below use the [httpie](https://httpie.io/) command line tool for making HTTP requests. You can also use curl or any other tool that can make HTTP requests.

`http get localhost:8000`

or

`curl -X GET http://localhost:8000/`

```http
HTTP/1.0 200 OK
Content-type: text/plain
Date: Fri, 13 Oct 2023 16:16:42 GMT
Server: SimpleHTTP/0.6 Python/3.11.6

Locationator server version 0.0.1 is running on port 8000
```

### GET /reverse_geocode

Receive geocode queries from the client. This endpoint accepts GET requests with latitude and longitude data as query parameters, performs reverse geocoding and returns the result.

**URL** : `/reverse_geocode?latitude=double&longitude=double`

**Method** : `GET`

**Query Parameters** :

|Parameter|Type|Description|
|---|---|---|
|`latitude`|Double|Latitude of the location to be reverse geocoded|
|`longitude`|Double|Longitude of the location to be reverse geocoded|

**Response format** :

- On Success, Content-type is application/json and a response code of 200 with a JSON object containing the reverse geocoding result is returned
- On Failure, a description of the error is returned with a suitable HTTP response code

**Success Response Example**:

`http GET "http://localhost:8000/reverse_geocode?latitude=33.953636&longitude=-118.338950"`

or

`curl -X GET "http://localhost:8000/reverse_geocode?latitude=33.953636&longitude=-118.338950"`

```http
HTTP/1.0 200 OK
Content-type: application/json
Date: Fri, 13 Oct 2023 19:32:24 GMT
Server: SimpleHTTP/0.6 Python/3.11.6

{
    "ISOcountryCode": "US",
    "administrativeArea": "CA",
    "areasOfInterest": [
        "SoFi Stadium"
    ],
    "country": "United States",
    "inlandWater": "",
    "locality": "Inglewood",
    "location": [
        33.953636,
        -118.33895
    ],
    "name": "SoFi Stadium",
    "ocean": "",
    "postalAddress": {
        "ISOCountryCode": "US",
        "city": "Inglewood",
        "country": "United States",
        "postalCode": "90305",
        "state": "CA",
        "street": "1001 Stadium Dr",
        "subAdministrativeArea": "Los Angeles County",
        "subLocality": "Century"
    },
    "postalCode": "90305",
    "subAdministrativeArea": "Los Angeles County",
    "subLocality": "Century",
    "subThoroughfare": "1001",
    "thoroughfare": "Stadium Dr",
    "timeZoneAbbreviation": "PDT",
    "timeZoneName": "America/Los_Angeles",
    "timeZoneSecondsFromGMT": -25200
}
```

## Services

Locationator will install two services accessible via the Services menu:

- `Locationator: reverse geocode`: extracts the latitude and longitude from the selected image and performs a reverse geocode lookup. The result is copied to the clipboard.
- `Locationator: write XMP data`: extracts the latitude and longitude from the selected image and writes the reverse geocode data to the XMP metadata of the image.

>*Note*: the `Locationator: write XMP data` service does not work with RAW image.

## Command Line Tools

Locationator provides a menu option to "Install Command Line tools" that will install the `locationator` CLI to `/usr/local/bin`. The CLI has 3 commands. `lookup`, `from-exif`, and `write-xmp`.  `lookup` looks up the reverse geocoding for a given latitude & longitude. `from-exif` reads the GPS coordinates from a file using [exiftool](https://exiftool.org) and prints out the reverse geocoding data in JSON format. `write-xmp` reads the GPS coordinates from the file and writes the following fields to the XMP metadata of the file using exiftool:

- XMP:CountryCode (ISOcountryCode)
- XMP:Country (country)
- XMP:State (administrativeArea)
- XMP:City (locality)
- XMP:Location (name)

exiftool must be installed to use `from-exif` or `write-xmp`.

The installation dialog will copy the required installation commands to the clipboard and prompt you to paste and run the commands in the terminal. You may be prompted for your admin password to complete the installation.

The following commands can be also run in the terminal to install the CLI:

```bash
[ ! -d "/usr/local/bin" ] && sudo mkdir -p /usr/local/bin
sudo ln -s /Applications/Locationator.app/Contents/Resources/locationator /usr/local/bin/locationator
```

To remove the CLI, run the following commands in the terminal:

```bash
sudo rm /usr/local/bin/locationator
```

CLI Usage:

```
Usage: locationator [OPTIONS] COMMAND [ARGS]...

Options:
  --debug
  --port INTEGER
  --help          Show this message and exit.

Commands:
  from-exif  Lookup the reverse geolocation for an image/video file using...
  lookup     Lookup the reverse geolocation for a lat/lon pair.
  write-xmp  Lookup the reverse geolocation for an image/video file using...
```

>*Note*: The CLI is a standalone binary created with [pyinstaller](https://pyinstaller.org/en/stable/) and the start-up time may be slow depending on your computer due to the way pyinstaller bundles the python interpreter and all dependencies into a single binary.

## Notes

- If building with [pyenv](https://github.com/pyenv/pyenv) installed python, you'll need to build the python with framework support:
  - `env PYTHON_CONFIGURE_OPTS="--enable-framework" pyenv install -v 3.11.6`
- Tested on macOS Ventura 13.5.1
- By default, the server runs on port 8000. This can be changed by editing the configuraiton plist file at `~/Library/Application Support/Locationator/Locationator.plist` and changing `port` to the desired port number then restarting the app.

## Building

Locationator uses [doit](https://pydoit.org/) to build the app. To build the app, run `doit` in the root of the repo. Run `doit list` to see list of available tasks.

- `pip install -r dev_requirements.txt`
- `doit`

Locationator has two components: a server which runs as a menu bar app and is built with [py2app](https://py2app.readthedocs.io/en/latest/), and a CLI which can be used to interact with the server and is built with [pyinstaller](https://pyinstaller.org/en/stable/).

The CLI must be built as a stand alone file with pyinstaller and the resulting `locationator` binary must be placed into the `locationator.app` bundle in the `Contents/resources` directory when the main app is built with py2app. The `dodo.py` script, which is run with `doit`, will do this for you.

## Testing

- The app can be tested by running the tests in the `tests` directory with pytest.
- `pip install -r dev_requirements.txt`
- `pytest -vv tests/`

>**Note**: the tests require the built app to be installed in /Applications

## License

Copyright (c) 2023, Rhet Turnbull  
Licensed under the MIT License


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.1.1](https://github.com/RhetTbull/locationator/compare/v0.1.0...v0.1.1)

> 28 October 2023

- Removed PUT API, #7 [`bb5959d`](https://github.com/RhetTbull/locationator/commit/bb5959def2ee79d7378baf156d8de179df6db08f)
- version bump [`41a383d`](https://github.com/RhetTbull/locationator/commit/41a383d5c8a4496d68bf9c2155bc98bbe519f98e)

#### [v0.1.0](https://github.com/RhetTbull/locationator/compare/v0.0.7...v0.1.0)

> 28 October 2023

- Service [`#11`](https://github.com/RhetTbull/locationator/pull/11)
- Initial Service working [`c5a6efb`](https://github.com/RhetTbull/locationator/commit/c5a6efbd25c2018b8668d63c9744b75d511f06f7)
- XMP service working [`350cb3a`](https://github.com/RhetTbull/locationator/commit/350cb3ae5672aaa54dee06cb2f9ac5de8e689ebe)
- Refined initial service [`11d5f03`](https://github.com/RhetTbull/locationator/commit/11d5f03daa8f1031fa4cf4cb2b1e28aca78d3d3d)
- Services working [`360c825`](https://github.com/RhetTbull/locationator/commit/360c825833e895e6125c63b1157166ee8bad095e)
- Fixed build script [`c13613d`](https://github.com/RhetTbull/locationator/commit/c13613d88e93413787ae20c75c7dc652e12f1c5c)

#### [v0.0.7](https://github.com/RhetTbull/locationator/compare/v0.0.6...v0.0.7)

> 21 October 2023

- Improved CLI install, #10 [`409f711`](https://github.com/RhetTbull/locationator/commit/409f71198ea23c3f745810e33edacf422a3f08ee)
- Added create_symbolic_link [`8688719`](https://github.com/RhetTbull/locationator/commit/8688719734e245d304f1e4100ca08d0015133fcf)
- Added installation note [`987df98`](https://github.com/RhetTbull/locationator/commit/987df98fa9e0e1f7e2219371505dd0d1912673a7)

#### [v0.0.6](https://github.com/RhetTbull/locationator/compare/v0.0.5...v0.0.6)

> 16 October 2023

- Refactored source files [`#9`](https://github.com/RhetTbull/locationator/pull/9)
- Update README.md [`8fcdacb`](https://github.com/RhetTbull/locationator/commit/8fcdacbe68c3362812f96c78347703284e0ffbb8)
- More refactoring [`564e75a`](https://github.com/RhetTbull/locationator/commit/564e75a922326d87fb2f7cdc36d0efe527e11ef2)
- Updated CLI doc strings [`6831735`](https://github.com/RhetTbull/locationator/commit/6831735026dea7c714b1f790b1b67da49fdb6051)
- Version bump [`5856ef3`](https://github.com/RhetTbull/locationator/commit/5856ef363262aaece8e33b132921970fc5ec90e4)
- Version bump [`402e889`](https://github.com/RhetTbull/locationator/commit/402e8899efdcf8eb766a56431998cd9bb6eabbae)

#### [v0.0.5](https://github.com/RhetTbull/locationator/compare/v0.0.4...v0.0.5)

> 16 October 2023

- Added write-xmp, from-exif to CLI [`cc53729`](https://github.com/RhetTbull/locationator/commit/cc53729df6f06c8e4c82e385308b33e84fb94d3f)
- Added help for negative values [`9b2bd42`](https://github.com/RhetTbull/locationator/commit/9b2bd425131fe5b317229ff7d63c014abd1ec2d8)

#### [v0.0.4](https://github.com/RhetTbull/locationator/compare/v0.0.3...v0.0.4)

> 15 October 2023

- Added initial CLI [`1b1dce3`](https://github.com/RhetTbull/locationator/commit/1b1dce3ec78ba0b5e12213b670f06bcbf7f8414b)
- Fixed debug bug, added dodo for building, added locationator CLI [`0c12aef`](https://github.com/RhetTbull/locationator/commit/0c12aefe59cdabd4acb16636bbb9a6f4d58c168f)
- Removed unneeded icon code [`5e0cd83`](https://github.com/RhetTbull/locationator/commit/5e0cd835c7a985db8ed0bcc6154b1499da394eb7)

#### [v0.0.3](https://github.com/RhetTbull/locationator/compare/v0.0.2...v0.0.3)

> 14 October 2023

- Added tests [`c7496ca`](https://github.com/RhetTbull/locationator/commit/c7496ca5f0a12c4d57d7ca63bda8cefdae12ed68)
- Refactored server code [`ef805b1`](https://github.com/RhetTbull/locationator/commit/ef805b11794f903ce5b6b8e6b7001b4644d3f9ac)
- Fixed reverse geocode feature in GUI [`177643b`](https://github.com/RhetTbull/locationator/commit/177643bf3ac05198092fd0eb80af84edd9fdf018)

#### [v0.0.2](https://github.com/RhetTbull/locationator/compare/v0.0.1...v0.0.2)

> 13 October 2023

- Added GET endpoint for reverse geocoding, #3 [`8bca085`](https://github.com/RhetTbull/locationator/commit/8bca0850de347e9e9344c050be8ac43d8f309ebc)
- Added curl example [`90f999d`](https://github.com/RhetTbull/locationator/commit/90f999db406f9f236423f603542cc00f63911bb3)
- Version bump [`6f3bea6`](https://github.com/RhetTbull/locationator/commit/6f3bea6da92417ca62cc6634e5af1fc239317d49)

#### v0.0.1

> 13 October 2023

- First commit [`3ad1f53`](https://github.com/RhetTbull/locationator/commit/3ad1f5352bc67f2f06b222206267545d4d2a7bc4)
- Cleaned up unneeded code [`d0e679b`](https://github.com/RhetTbull/locationator/commit/d0e679ba89ee8da6aa8c22023a0ad4757eccfac3)
- Doesn't crash [`1cb744f`](https://github.com/RhetTbull/locationator/commit/1cb744fc2e3817d1a892eb76cdb290568d2f1933)


locationator/cli.py
"""CLI for Locationator to get reverse geolocation for a lat/lon pair."""

from __future__ import annotations

import json
import pathlib
import plistlib
from typing import Any

import click
import httpx
from exiftool import ExifTool, get_exiftool_path

# Do not update __version__ manually. Use bump2version.
__version__ = "0.1.1"


@click.group()
@click.option("--debug", is_flag=True, help="Enable debug mode")
@click.option(
    "--port",
    default=0,
    type=int,
    help="Locationator server port. Defaults to port in Locationator plist config file.",
)
@click.pass_context
def cli(ctx: click.Context, debug: bool, port: int):
    ctx.ensure_object(dict)
    ctx.obj["DEBUG"] = debug
    ctx.obj["PORT"] = port


@cli.command()
@click.pass_context
@click.option("--indent", "-i", type=int, help="Indentation level for JSON output")
@click.option("--no-indent", "-I", is_flag=True, help="Do not indent JSON output")
@click.argument("latitude", type=float)
@click.argument("longitude", type=float)
def lookup(
    ctx: click.Context, indent: int, no_indent: bool, latitude: float, longitude: float
):
    """Lookup the reverse geolocation for a lat/lon pair.

    Note: if you need to pass a negative latitude or longitude, you must use the -- flag to
    prevent the command line parser from interpreting the value as an option.
    This is the standard behavior for most command line tools.

    For example:

    locationator lookup -- 33.953636 -118.33895

    """
    port = get_port(ctx)
    indent = get_indent(indent, no_indent)

    results = reverse_geocode(latitude, longitude, port)
    if results:
        click.echo(json.dumps(results, indent=indent))
    else:
        click.echo("Error: Could not get reverse geolocation", err=True)
        raise click.Abort()


@cli.command(name="from-exif")
@click.option("--indent", "-i", type=int, help="Indentation level for JSON output")
@click.option("--no-indent", "-I", is_flag=True, help="Do not indent JSON output")
@click.argument("filename", type=click.Path(exists=True, dir_okay=False))
@click.pass_context
def from_exif(ctx: click.Context, indent: int, no_indent: bool, filename: str):
    """Lookup the reverse geolocation for an image/video file using the
    latitude/longitude from the file's metadata.

    Requires exiftool (https://exiftool.org/) be installed and on your PATH.

    """

    try:
        get_exiftool_path()
    except FileNotFoundError as e:
        click.echo(e, err=True)
        raise click.Abort()

    port = get_port(ctx)
    indent = get_indent(indent, no_indent)

    exiftool = ExifTool(filename)
    metadata = exiftool.asdict()

    latitude = metadata.get("Composite:GPSLatitude")
    longitude = metadata.get("Composite:GPSLongitude")

    if latitude is None or longitude is None:
        click.echo("Error: Could not find GPS latitude/longitude in metadata", err=True)
        raise click.Abort()

    latitude = float(latitude)
    longitude = float(longitude)

    results = reverse_geocode(latitude, longitude, port)
    if results:
        click.echo(json.dumps(results, indent=indent))
    else:
        click.echo("Error: Could not get reverse geolocation", err=True)
        raise click.Abort()


@cli.command(name="write-xmp")
@click.option("--indent", "-i", type=int, help="Indentation level for JSON output")
@click.option("--no-indent", "-I", is_flag=True, help="Do not indent JSON output")
@click.argument("filename", type=click.Path(exists=True, dir_okay=False))
@click.pass_context
def write_xmp(ctx: click.Context, indent: int, no_indent: bool, filename: str):
    """Write the reverse geocode results to the file's XMP metadata after performing
    a lookup with the latitude/longitude found in the file's metadata with exiftool.

    Requires exiftool (https://exiftool.org/) be installed and on your PATH.

    The following XMP fields are written (parentheses indicate the corresponding
    reverse geocode result field):

    \b
    - XMP:CountryCode (ISOcountryCode)
    - XMP:Country (country)
    - XMP:State (administrativeArea)
    - XMP:City (locality)
    - XMP:Location (name)

    """

    try:
        get_exiftool_path()
    except FileNotFoundError as e:
        click.echo(e, err=True)
        raise click.Abort("Error: Could not find exiftool")

    port = get_port(ctx)
    indent = get_indent(indent, no_indent)

    exiftool = ExifTool(filename)
    metadata = exiftool.asdict()

    latitude = metadata.get("Composite:GPSLatitude")
    longitude = metadata.get("Composite:GPSLongitude")

    if latitude is None or longitude is None:
        click.echo("Error: Could not find GPS latitude/longitude in metadata", err=True)
        raise click.Abort()

    latitude = float(latitude)
    longitude = float(longitude)

    results = reverse_geocode(latitude, longitude, port)
    if not results:
        click.echo("Error: Could not get reverse geolocation", err=True)
        raise click.Abort()

    xmp = write_xmp_metadata(filename, results)
    click.echo(f"Wrote the following XMP metadata to {filename}:")
    for key, value in xmp.items():
        click.echo(f"{key}: {value}")


def reverse_geocode(latitude: float, longitude: float, port: int) -> dict[str, Any]:
    """Perform reverse geocode of latitude/longitude

    Args:
        latitude (float): Latitude
        longitude (float): Longitude
        port (int): Port number to connect to

    Returns: dict: Reverse geocode result
    """
    with httpx.Client() as client:
        try:
            response = client.get(
                f"http://localhost:{port}/reverse_geocode?latitude={latitude}&longitude={longitude}"
            )
            if response.status_code == 200:
                return response.json()
            else:
                click.echo(f"Error: {response.status_code} {response.text}", err=True)
                return {}
        except httpx.ConnectError:
            click.echo(
                f"Error: Could not connect to server on port {port}. Is Locationator.app running?",
                err=True,
            )
            return {}


def write_xmp_metadata(filename: str, results: dict[str, Any]) -> dict[str, Any]:
    """Write reverse geolocation-related fields to file metadata using exiftool

    Args:
        filename (str): Path to file
        results (dict): Reverse geocode results

    Note: The following XMP fields are written (parentheses indicate the corresponding
    reverse geocode result field):

    - XMP:CountryCode (ISOcountryCode)
    - XMP:Country (country)
    - XMP:State (administrativeArea)
    - XMP:City (locality)
    - XMP:Location (name)
    """

    metadata = {
        "XMP:CountryCode": results["ISOcountryCode"],
        "XMP:Country": results["country"],
        "XMP:State": results["administrativeArea"],
        "XMP:City": results["locality"],
        "XMP:Location": results["name"],
        # "XMP-iptcExt:LocationCreatedSublocation": results["subLocality"],
    }

    with ExifTool(filename) as exiftool:
        for key, value in metadata.items():
            exiftool.setvalue(key, value)

    return metadata


def load_config() -> dict[str, Any]:
    """Load config from the Locationator plist file"""
    plist_path = pathlib.Path(
        "~/Library/Application Support/Locationator/Locationator.plist"
    ).expanduser()
    if not plist_path.exists():
        raise FileNotFoundError(
            f"Could not find Locationator plist file at {plist_path}"
        )
    with open(plist_path, "rb") as f:
        return plistlib.load(f)


def get_port(ctx: click.Context) -> int:
    """Get the port number to connect to from command line or config file"""
    try:
        config = load_config()
    except FileNotFoundError as e:
        click.echo(e, err=True)
        config = {}

    return ctx.obj["PORT"] if ctx.obj["PORT"] else config.get("port", 8000)


def get_indent(indent: int, no_indent: bool) -> int | None:
    """Return value for json indent argument"""
    if no_indent and indent is not None:
        raise click.UsageError("Cannot specify both --indent and --no-indent")

    if no_indent:
        return None

    return indent if indent is not None else 4


if __name__ == "__main__":
    cli(obj={})
